File: tiq-e/dynamic-programming/2/main.go
package main

import (
	"fmt"
	"math"
	"time"
)

func maxProfit(prices []int) int {
	minprice := math.MaxInt32
	maxprofit := 0
	for _, price := range prices {
		if price < minprice {
			minprice = price
		} else if price-minprice > maxprofit {
			maxprofit = price - minprice
		}
	}
	return maxprofit
}

func main() {
	var n = []int{7, 1, 5, 3, 6, 4}

	var start = time.Now()
	var result = maxProfit(n)
	var end = time.Since(start)
	fmt.Println("Duration: ", end)
	fmt.Println("Result: ", result)
}

File: tiq-e/dynamic-programming/2/main.ts
import {performance} from "perf_hooks";

function maxProfit(prices: number[]): number {
    let minprice = Infinity
    let maxprofit = 0
    for (let i = 0; i < prices.length; i++) {
        if (prices[i] < minprice) minprice = prices[i]
        else if (prices[i] - minprice > maxprofit) maxprofit = prices[i] - minprice
    }

    return maxprofit

}

const n = [7, 1, 5, 3, 6, 4];

const start = performance.now();
const result = maxProfit(n);
const end = performance.now();
console.log("Duration:", end - start);
console.log("Result:", result);
